### PR TITLE
[MAMA] thread-safe version of endpoint pool

### PR DESCRIPTION
**Pull request by bill-torpey-ullink**
## Summary
thread-safe version of endpoint pool

## Areas Affected
*Place an 'x' within the braces to check the box*
- [x] MAMAC
- [ ] MAMACPP
- [ ] MAMADOTNET
- [ ] MAMAJNI
- [ ] MAMDA
- [ ] MAMDACPP
- [ ] MAMDADOTNET
- [ ] MAMDAJNI
- [ ] Visual Studio
- [ ] SCons
- [ ] Unit Tests
- [ ] Examples

## Details
Thread-safe version of endpoint pool -- see https://github.com/fquinner/OpenMAMA-zmq/issues/19

## Testing
With this patch, the zmq transport no longer cores under load.

### CODE REVIEW

File: mama/c_cpp/src/c/bridge/qpid/endpointpool.c [hunk 1]
```diff
@@ -428,21 +450,28 @@ endpointPool_isRegistedByContent (endpointPool_t    endpoints,
         return MAMA_STATUS_NULL_ARG;
     }
 
+    wlock_lock(impl->mLock);
+
     /* Get the wtable representing the endpoints associated with this topic */
+    int ret;
```
Comment (fquinner) [inline mama/c_cpp/src/c/bridge/qpid/endpointpool.c]: Mixed variable here otherwise it looks good.

**Comment by bill-torpey-ullink** [comment]
I assume by "mixed" you mean that ret is declared non-K&R style (i.e., near point of first use)?  My bad -- the coding standards do state that K&R style is mandated, so I've changed that.  This is a C89 standard, so maybe it would be best to describe the convention that way?

FWIW, VS 2013 apparently supports mixing declarations and code (https://msdn.microsoft.com/en-us/library/hh409293(v=vs.120).aspx), so unless OpenMAMA is targeting earlier versions, it may be worth revisiting this restriction.  

Assuming that the C89 standard is actually the point here, it should be possible to move the declaration of existNode into the body of the `if` statement.  This will also make static analysis tools like cppcheck happy (the existing declaration will trigger a scope warning in cppcheck).  I've done that in the latest PR, but let me know if that is a problem.  

Last but not least, further inspection reveals what appear to be a couple of real bugs in the original code:

- Returning MAMA_STATUS_NULL_ARG is effectively returning true, which doesn't appear to be the intention (i.e., zero means 'not found', anything else means 'found').
- The code in the body of the `if` should return the result from `endpointPoolImpl_checkEndpointExists` (which sets the mResult member to 1 only if there is a match), rather than simply returning 1 in all cases.

**Comment by bill-torpey-ullink** [comment]
One more thing re: "mixed" decls -- I notice that the Linux build specifically calls for `--std=gnu99`.  Isn't that incompatible with `--std=c89`?  Shouldn't the project be consistent wrt required compiler support?

**Comment by fquinner** [comment]
We never claimed to be C89 - check out https://openmama.github.io/openmama_coding_standards.html#portability - its more C99 plus whatever keeps older Microsoft compilers happy. And yes we still support them for now - see https://openmama.github.io/openmama_supported_platforms.html#windows. This is a PR though - mailing lists would be a better place to bring something like conforming to a specific C standard up.

**Comment by bill-torpey-ullink** [comment]
Well, it keeps coming up, but you're right.  

So I *think* the current PR is kosher as per standards, and believe it fixes the two addl. bugs mentioned above.  Please let me know.

**Comment by fquinner** [comment]
Hi Bill, looks like some conflicts have sneaked in - can you have a look please?

**Comment by bill-torpey-ullink** [comment]
The conflicts are related to the earlier PR -- https://github.com/OpenMAMA/OpenMAMA/pull/307.

There's probably a clever way to resolve this using git, but my git-fu is not that great.  If you have a suggestion on how to resolve this using git, let me know and I'll give it a go.

Otherwise I can just update the PR with the code from earlier PR.

**Comment by fquinner** [comment]
Yeah if you switch to the PR branch and do a `git pull --rebase origin next` (if origin is up to date), it'll merge and ask you to resolve conflicts as necessary. Don't worry if there are merge commits - I'll squash them when merging into next anyway.

**Comment by bill-torpey-ullink** [comment]
Well, git didn't help much -- just left a bunch of conflicting lines that I would have thought (hoped?) it could merge on its own.  But BeyondCompare (scootersoftware.com) to the rescue!

Should be good to go now -- pls let me know.

**Pull request merged**